src/pybind/mgr/dashboard/services/rbd.py

```diff
@@ -662,8 +662,8 @@ def _copy(d_ioctx):
         return rbd_image_call(pool_name, namespace, image_name, _src_copy)
 
     @classmethod
+    @ttl_cache_invalidator(CacheReference('rbd_image_refs'))
```

Comment:
I went for whole cache invalidation in rbd for now for simplicity. Going for specific key invalidation will mean extra logic on top of the wrapper or adding inline invalidation plus ensuring the construction of the key is the  same in both the cached function and in invalidation.

Reply:
There are a couple of caveats here:
* The explicit cache invalidation will only catch the changes from the Dashboard (through the REST API). I'm almost sure that there should be another mechanism in `librbd` to detect that (maybe an increasing versioning number or some hash)... Different story is whether that's available in a public API.
* That said, I still see that this explicit cache invalidation is a bit coarse-grained and reduces the overall efficiency of the cache.

Here we have 3 cached class methods (meaning that the first argument will be the same):
* `get_ioctx(pool_name, namespace)`
* `_pool_namespaces(pool_name, namespace=None)`
* `_rbd_image_refs(pool_name, namespace='')`

That means that we'll have 3 procedural caches with 3 keys:
* ioctx cache `(pool, namespace)`, to improve RADOS connection times,
* pool cache `(pool, namespace)`, to improve pool connection times,
* rbd cache `(pool, namespace)`, to improve listing of all RBD images.

For a fine-grained invalidatable cache, we should move away from a procedural cache and use a map-cache (which funnily is what lies beneath the procedural interface). So basically we keep a `TTLRBDImageCache[(pool_name, image_name)]`. All RBDService methods actually read and populate from that cache.

* `create` creates a new RBD image and `get()`s it from RBD (in order to ensure that the transaction was successful and that we're reading the actual data from RBD and not _what we think we told RBD to store_). We populate the `TTLRBDImageCache` with this data.
* `delete` deletes the RBD image and ensures (via return code or get) that the image is not there anymore.
* Same for `set`, `copy` and `flatten`...

With this approach we could have a separate thread (don't like this too much) ensuring that the TTLCache is consistent with the RBD data. In fact, if we have such a background-sync'ing thread, we could probably need no TTL at all, since the data is eventually consistent, and the syncing-period replaces the expiration period.

What do you think?

Reply:
So basically you want a caching layer that works somewhat like the view cache.

I understand the issue with invalidating the cache based on dashboard operations because cli operations are ignored when proceeding to invalidate. 
One issue that I can think about is book keeping pools, we can either get the pools from the input of `rbd_pool_list` and update the `TTLRBDImageCache` so that it knows whichs pools to keep track of, or we call osd_map periodically to check pools with rbd application on it. This could be a great time to add fine grained get_pools api in the mgr also :)

Reply:
> So basically you want a caching layer that works somewhat like the view cache.

I love it, I hate it... 🙈 If the image listing is performant enough to work synchronously, then we could ignore that background thread that the ViewCache had. If not, we either add a background async thread, or hard-limit the number of images (e.g.: `LIMIT=100,000`, since `rbd` provides an iterator based `list` method).

But if I recall it correctly the big issue happened with the `stat`-like method iterating over N images, right? So we can keep the listing synchronous, and perform the rbd image stat on demand too.

Having a fine-grained API is a good thing, as long as the scalability issue lies in there, does it?

Reply:
Even though it is true invalidation detection from ceph core is not present with this code, I've decided to skip it for now as this code includes some useful infra for host listing improvements we need right now. 

I could say almost 100% for sure listing image references is fast on the development evironment but for now I cannot prove it otherwise on larger clusters as I'm not sure how rbd stores this data.